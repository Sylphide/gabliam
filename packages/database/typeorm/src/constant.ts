export const ConnectionOptionsBeanId = Symbol('ConnectionOptionsBeanId');

export const TYPE = {
  Entity: 'EntityTypeOrm',
};

export const METADATA_KEY = {
  cunit: '_cunit',
};

export const ENTITIES_TYPEORM = Symbol('TYPEORM/ENTITIES_TYPEORM');
