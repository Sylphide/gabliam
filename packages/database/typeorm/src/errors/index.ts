export * from './typeorm-config-is-mandatory-error';
export * from './typeorm-connection-not-found-error';
export * from './typeorm-cunit-not-found-error';
