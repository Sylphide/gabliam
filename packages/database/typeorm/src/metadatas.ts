import {
  Entity as typeormEntity,
  ChildEntity as typeormChildEntity,
  EntityOptions,
} from 'typeorm';
import { Register, makeDecorator } from '@gabliam/core';
import { TYPE, METADATA_KEY, ERRORS_MSGS } from './constant';

/**
 * Type of the `CUnit` decorator / constructor function.
 */
export interface CUnitDecorator {
  /**
   * Decorator that marks a class (a typeorm Entity) to use a specific connection
   *
   * @usageNotes
   *
   * ```typescript
   * @CUnit('test')
   * @Entity()
   * export class Photo {
   *   @PrimaryGeneratedColumn() id: number;
   *
   *   @Column({
   *     length: 500
   *   })
   *   name: string;
   *
   *   @Column('text') description: string;
   *
   *   @Column() fileName: string;
   *
   *   @Column('int') views: number;
   *
   *   @Column() isPublished: boolean;
   * }
   * ```
   */
  (name: string): ClassDecorator;

  /**
   * see the `@CUnit` decorator.
   */
  new (name: string): any;
}

/**
 * `CUnit` decorator and metadata.
 */
export interface CUnit {
  /**
   * Name of the connection
   */
  name: string;
}

export const CUnit: CUnitDecorator = makeDecorator(
  METADATA_KEY.cunit,
  (name: string): CUnit => ({ name }),
  undefined,
  true,
  ERRORS_MSGS.DUPLICATED_CUNIT_DECORATOR
);

/**
 * Type of the `Entity` decorator / constructor function.
 */
export interface EntityDecorator {
  /**
   * This decorator is used to mark classes that will be an entity
   * (table or document depend on database type).
   * Database schema will be created for all classes decorated with it,
   * and Repository can be retrieved and used for it.
   *
   * @usageNotes
   *
   * ```typescript
   * @Entity()
   * export class Photo {
   *   @PrimaryGeneratedColumn() id: number;
   *
   *   @Column({
   *     length: 500
   *   })
   *   name: string;
   *
   *   @Column('text') description: string;
   *
   *   @Column() fileName: string;
   *
   *   @Column('int') views: number;
   *
   *   @Column() isPublished: boolean;
   * }
   * ```
   */
  (name?: string, options?: EntityOptions): ClassDecorator;

  /**
   * see the `@Entity` decorator.
   */
  new (name?: string, options?: EntityOptions): any;
}

/**
 * `Entity` decorator and metadata.
 */
export interface Entity {
  name?: string;

  options?: EntityOptions;
}

export const Entity: EntityDecorator = makeDecorator(
  'Entity',
  (name?: string, options?: EntityOptions): Entity => ({ name, options }),
  (cls, annotationInstance: Entity) => {
    Register({ type: TYPE.Entity, id: cls, autobind: false })(cls);
    typeormEntity(annotationInstance.name, annotationInstance.options)(cls);
  }
);

/**
 * Type of the `ChildEntity` decorator / constructor function.
 */
export interface ChildEntityDecorator {
  /**
   * Special type of the table used in the single-table inherited tables.
   *
   * @usageNotes
   *
   * ```typescript
   * @Entity()
   * @TableInheritance({ column: { type: "varchar", name: "type" } })
   * export class Content {
   *
   *     @PrimaryGeneratedColumn()
   *     id: number;
   *
   *     @Column()
   *     title: string;
   *
   *     @Column()
   *     description: string;
   *
   * }
   *
   * @ChildEntity()
   * export class Photo extends Content {
   *
   *     @Column()
   *     size: string;
   *
   * }
   * ```
   */
  (): ClassDecorator;

  /**
   * see the `@ChildEntity` decorator.
   */
  new (): any;
}

export const ChildEntity: ChildEntityDecorator = makeDecorator(
  'ChildEntity',
  undefined,
  cls => {
    Register({ type: TYPE.Entity, id: cls, autobind: false })(cls);
    typeormChildEntity()(cls);
  }
);
