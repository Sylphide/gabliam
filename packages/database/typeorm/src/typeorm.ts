export * from 'typeorm';

import { Entity, ChildEntity, CUnit } from './metadatas';

export { Entity, ChildEntity, CUnit };
