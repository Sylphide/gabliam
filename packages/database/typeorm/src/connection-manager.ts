import { reflection, ValueExtractor } from '@gabliam/core';
import { createConnections } from 'typeorm';
import {
  TypeormConnectionNotFoundError,
  TypeormCUnitNotFoundError,
} from './errors';
import { CUnit } from './metadatas';
import { Connection, ConnectionOptions } from './typeorm';

export class ConnectionManager {
  private connections!: Connection[];

  constructor(
    private connectionOptions: ConnectionOptions[],
    private entities: Function[],
    private valueExtractor: ValueExtractor
  ) {}

  async open() {
    const connectionOptions = this.connectionOptions.map(c => {
      return {
        ...c,
        entities: Array.isArray(c.entities) ? c.entities : [],
        name: c.name ? c.name : 'default',
      };
    });

    // add entity to the correct connection
    for (const entity of this.entities) {
      const cunits = getCunits(entity);

      for (let cunit of cunits) {
        cunit = this.valueExtractor(cunit, cunit);
        let index = connectionOptions.findIndex(c => c.name === cunit);

        if (index === -1 && cunit === 'default') {
          index = 0;
        }

        if (index === -1) {
          throw new TypeormCUnitNotFoundError(cunit);
        }
        (<any>connectionOptions)[index].entities.push(entity);
      }
    }

    this.connections = await createConnections(connectionOptions);
  }

  async close() {
    await Promise.all(this.connections.map(c => c.close()));
  }

  getConnection(name: string) {
    const connection = this.connections.find(c => c.name === name);
    if (!connection) {
      throw new TypeormConnectionNotFoundError(name);
    }
    return connection;
  }

  getDefaultConnection() {
    const connection = this.connections.find(c => c.name === 'default');
    if (connection === undefined) {
      return this.connections[0];
    } else {
      return connection;
    }
  }
}

const getCunits = (cls: any) => {
  const cunits = reflection.annotationsOfDecorator<CUnit>(cls, CUnit);

  if (cunits.length) {
    return cunits.map(c => c.name || 'default');
  }

  return ['default'];
};
