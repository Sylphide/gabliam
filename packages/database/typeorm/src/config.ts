import {
  Bean,
  Container,
  Init,
  inject,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  optional,
  PluginConfig,
  Value,
  ValueExtractor,
  VALUE_EXTRACTOR,
} from '@gabliam/core';
import * as d from 'debug';
import { ConnectionManager } from './connection-manager';
import { ConnectionOptionsBeanId, ENTITIES_TYPEORM } from './constant';
import { TypeormConfigIsMandatoryError } from './errors';
import { Connection, ConnectionOptions } from './typeorm';

const debug = d('Gabliam:Plugin:Typeorm');

@InjectContainer()
@PluginConfig()
export class PluginTypeormConfig {
  @Value('application.typeorm.connectionOptions')
  connectionOptions: ConnectionOptions | ConnectionOptions[];

  entities: Function[];

  constructor(
    @inject(ConnectionOptionsBeanId)
    @optional()
    connectionOptions: ConnectionOptions | ConnectionOptions[],
    @inject(ENTITIES_TYPEORM) entities: Function[]
  ) {
    debug('constructor PluginTypeormConfig', connectionOptions);
    this.connectionOptions = connectionOptions;
    this.entities = entities;
  }

  // when all bean are created, we create bean Connection for back compat
  @Init()
  async init() {
    const container: Container = (<any>this)[INJECT_CONTAINER_KEY];
    const connectionManager = container.get(ConnectionManager);
    await connectionManager.open();

    // for back compat
    container
      .bind(Connection)
      .toConstantValue(connectionManager.getDefaultConnection());
  }

  @Bean(ConnectionManager)
  createManager() {
    debug('connectionOptions', this.connectionOptions);
    if (!this.connectionOptions) {
      throw new TypeormConfigIsMandatoryError();
    }
    let connectionOptions: ConnectionOptions[];
    if (Array.isArray(this.connectionOptions)) {
      connectionOptions = this.connectionOptions;
    } else {
      connectionOptions = [this.connectionOptions];
    }

    const container: Container = (<any>this)[INJECT_CONTAINER_KEY];
    const valueExtractor = container.get<ValueExtractor>(VALUE_EXTRACTOR);

    return new ConnectionManager(
      connectionOptions,
      this.entities,
      valueExtractor
    );
  }
}
