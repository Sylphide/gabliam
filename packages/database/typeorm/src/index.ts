export * from './typeorm';
export * from './constant';
export { ConnectionManager } from './connection-manager';
import { TypeOrmPlugin } from './typeorm-plugin';
export { TypeOrmPlugin as default };
