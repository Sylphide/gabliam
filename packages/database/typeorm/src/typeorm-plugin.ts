import {
  Scan,
  Registry,
  Plugin,
  GabliamPlugin,
  ValueRegistry,
  Container,
} from '@gabliam/core';
import { TYPE, ENTITIES_TYPEORM } from './constant';
import { ConnectionManager } from './connection-manager';
import { ContainerInterface, ContainedType, useContainer } from './typeorm';

/**
 * Container to be used by this library for inversion control. If container was not implicitly set then by default
 * container simply creates a new instance of the given class.
 */
const defaultContainer: ContainerInterface = new class
  implements ContainerInterface {
  public instances: { type: Function; object: any }[] = [];

  get<T>(someClass: ContainedType<T>): T {
    let instance = this.instances.find(i => i.type === someClass);
    if (!instance) {
      instance = { type: someClass, object: new (someClass as new () => T)() };
      this.instances.push(instance);
    }

    return instance.object;
  }
}();

useContainer(defaultContainer);

@Plugin()
@Scan()
export class TypeOrmPlugin implements GabliamPlugin {
  bind(container: Container, registry: Registry) {
    const entities = registry
      .get<ValueRegistry>(TYPE.Entity)
      .map(({ target }) => {
        return target;
      });
    container.bind<any>(ENTITIES_TYPEORM).toConstantValue(entities);
  }

  async destroy(container: Container, registry: Registry) {
    const connection = container.get(ConnectionManager);
    await connection.close();
    (<any>defaultContainer).instances = [];
  }
}
