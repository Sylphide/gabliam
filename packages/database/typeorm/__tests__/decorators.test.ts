import { reflection } from '@gabliam/core';
import { ChildEntity, CUnit, Entity } from '../src';

[Entity, ChildEntity].forEach(deco => {
  test(`${deco.name} decorators`, () => {
    @deco()
    class TestEntity {}

    expect(reflection.annotations(TestEntity)).toMatchSnapshot();
  });
});

describe('CUnit decorators', () => {
  test('should add CUnit metadata to a class when decorating a method with @CUnit', () => {
    @CUnit('default')
    class TestEntity {}

    expect(reflection.annotations(TestEntity)).toMatchSnapshot();
  });

  test('should add CUnit metadata to a class when decorating multiple times with @CUnit', () => {
    @CUnit('default')
    @CUnit('default2')
    class TestEntity {}

    expect(reflection.annotations(TestEntity)).toMatchSnapshot();
  });
});
