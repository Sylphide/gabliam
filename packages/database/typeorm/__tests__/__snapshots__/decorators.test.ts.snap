// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CUnit decorators should add CUnit metadata to a class when decorating a method with @CUnit 1`] = `
Array [
  DecoratorFactory {
    "name": "default",
  },
]
`;

exports[`CUnit decorators should add CUnit metadata to a class when decorating multiple times with @CUnit 1`] = `
Array [
  DecoratorFactory {
    "name": "default2",
  },
  DecoratorFactory {
    "name": "default",
  },
]
`;

exports[`DecoratorFactory decorators 1`] = `
Array [
  DecoratorFactory {
    "name": undefined,
    "options": undefined,
  },
  DecoratorFactory {
    "autobind": false,
    "id": [Function],
    "type": "EntityTypeOrm",
  },
]
`;

exports[`DecoratorFactory decorators 2`] = `
Array [
  DecoratorFactory {},
  DecoratorFactory {
    "autobind": false,
    "id": [Function],
    "type": "EntityTypeOrm",
  },
]
`;
